Nextjs

kristojorg opened the pull request:
This PR is a refactor of the build system, server code, routing, and some other smaller bits to move the app over to Next.js. The goal here was to reduce code surface area/complexity, and to set us up for future improvements based on Next.js. Things to note:

1. We now use file-system based routing instead of react router. The routes are all the components within `/pages`. The one complication is that both multi-library routes and single-library routes are always available no matter which setting we run the app with. @M0nica added checks to make sure the app returned 404 in the case we try to access a multi library route when running with a single library, or vice versa. 
2. Links work slightly different with Next.js. [Read here](https://nextjs.org/docs/api-reference/next/link). The basics are that Next.js doesn't ship a route manifest like react router does. This means that we have to tell the link both where in the `/pages` directory to route, and what url to display in the browser. To make things simpler, I refactored the `pathFor` function into `LinkUtils`, which has three functions to help with build route for books and collections. We can potentially remove the `PathForProvider` when re-examining the relationship with opds-web-client.

In the future, we will be moving data fetching to take advantage of the Next.js APIs, which will greatly improve performance and continue to simplify the app.

EdwinGuzman commented:
I'm sort of swamped with my current task so this might take some time, if that's okay. I'll start reviewing over the weekend.

kristojorg commented:
@EdwinGuzman no problem. I'll just base future work off this branch

EdwinGuzman commented on .gitignore:
```diff
@@ -1,10 +1,35 @@
-npm-debug.log
+# dependencies
+/node_modules
+/.pnp
+.pnp.js
```
What's this for?

kristojorg replied:
Not sure actually. This was generated by next.js originally. I'll remove it because I don't believe we need it.

EdwinGuzman commented on .vscode/tasks.json:
```diff
@@ -0,0 +1,12 @@
+{
```
Did you mean to add this? If so, how would I use it in vscode?

kristojorg replied:
Nope, not sure how this got there. Thanks for catching, have removed it.

EdwinGuzman commented on package.json:
```diff
@@ -9,84 +9,17 @@
+    "lint": "eslint '*/**/*.{ts,tsx}'",
+    "lint:fix": "eslint '*/**/*.{js,ts,tsx}' --quiet --fix",
+    "generate-icons": "npx @svgr/cli --template src/icons/_template.js --typescript --icon --ext tsx -d src/icons src/icons",
+    "storybook": "start-storybook -p 6006",
+    "build-storybook": "build-storybook"
```
I don't think you meant to add these, right?

kristojorg replied:
Ah, right we removed storybook in the beta PR. Thanks

EdwinGuzman commented on next-env.d.ts:
```diff
@@ -0,0 +1,2 @@
+/// <reference types="next" />
```
Is this for setting up the type for next to use typescript?

kristojorg replied:
Yep exactly. This file has to be there.

EdwinGuzman commented on src/components/Link.tsx:
```diff
@@ -1,46 +1,70 @@
+  }
+  if ("collectionUrl" in props) {
+    const { collectionUrl, ...rest } = props;
+    return { ...linkUtils.buildCollectionLink(props.collectionUrl), ...rest };
```
Minor, you don't need `props.collectionUrl` since you assigned it to a variable already.

EdwinGuzman commented on src/components/__tests__/Button.test.tsx:
```diff
@@ -28,13 +27,10 @@ test("disabled style", () => {
   const button = node.getByText("child");
   expect(button).toHaveAttribute("href", "/somewhere");
   expect(button).toMatchSnapshot();
-  // make sure clicking it actually tries to route you home
```
Is there an alternative for this test or why is there no need to test this anymore?

kristojorg replied:
Instead of testing that the click goes to the right place, we are testing that it has the right href, which does the same thing and is a more isolated test. I was running in to issues using next.js Link and jsdom which brought this up, and it became apparent that by testing the click I was testing the internals of these packages when what I really care about is the dom structure here.

kristojorg replied:
Just FYI testing the click functionality would still be possible if it were an actual button, but it's an internal link that just looks like a button which is why we test it using the href instead.

EdwinGuzman commented on src/components/__tests__/Error.test.tsx:
```diff
@@ -0,0 +1,35 @@
+import { render } from "../../test-utils";
+import Error from "../Error";
+
+test("renders error message without 404 status code if status code not provided", () => {
```
I think this should be `renders error message WITH 404 status code...`

EdwinGuzman commented on src/components/__tests__/Error.test.tsx:
```diff
@@ -0,0 +1,35 @@
+
+test("renders 404 error message when there is a 404 error", () => {
+  const node = render(<Error statusCode={404} />);
+  expect(
+    node.getByText("Error: This page could not be found")
```
This can also go in the previous test since it does render this by default.

EdwinGuzman commented on src/dataflow/getLibraryData.ts:
```diff
@@ -0,0 +1,153 @@
+}
+
+// the cache should only be created once per server instance.
+const cachePromise = setupCache();
```
Don't you need an `await` here?

kristojorg replied:
No because I'm actually trying to capture the promise here, not the final result yet. I await it later in `fetchLibraryData`, but wanted to create it at the top level because it should only ever be created once.

EdwinGuzman commented on src/pages/book/[bookUrl].tsx:
```diff
@@ -0,0 +1,13 @@
+
+const BookPage = () => {
+  return (
+    <Layout>
+      <BookDetails />
+    </Layout>
```
Why do these components not have the status code check to render the error page?

kristojorg replied:
This of how the routing works in Next.js when there are two different routing modes that are laid on top of each other (single library and multi library). If it is running as a single library and you try to visit `/something`, the app will think you are trying to see a multi library page, and we need to check if the app is running in multi library mode or not in order to 404 you appropriately. This doesn't happen on the single library routes, so we don't need the check there.

I do hope in the future there can be a different solution for multiple libraries without adding tons of devops overhead. I have a few ideas. Hopefully one of them can remove all of these edge cases and extra code from having two separate routing modes

EdwinGuzman commented on README.md:
```diff
@@ -76,6 +76,11 @@ The following environment variables can also be set to further configure the app
 - Set `SHORTEN_URLS=false` to stop the app from removing common parts of the circulation manager URLs from the web app's URLs.
```
The readme currently says to run the production build as `npm run build:prod` but it should be `npm run build`.

kristojorg replied:
Thanks!

EdwinGuzman commented:
Code looks good but here are some errors I see when running the app:

1) Using `REGISTRY_BASE=http://localhost:7000 npm run dev`:

The URL I go to is something like `http://localhost:3000/4d391d23-7bcc-45d3-9fcb-f7239a87a16c` where it's the uuid after the domain name (this is obtained from the library registry UI). I'm not sure why it's picking up the `cm_libraries` file even though I explicitly set the registry base. Everything still works but these errors are shown in the console. 

<img width="1022" alt="Screen Shot 2020-05-26 at 12 15 26 PM" src="https://user-images.githubusercontent.com/1280564/82924690-b3227e80-9f4a-11ea-851c-701b962f5402.png">

Let's say I decide to click on a book:
<img width="719" alt="Screen Shot 2020-05-26 at 12 18 08 PM" src="https://user-images.githubusercontent.com/1280564/82924927-098fbd00-9f4b-11ea-830d-b2d6e812854b.png">

And then click on "Edwin" in the breadcrumbs, I get this error:
<img width="1206" alt="Screen Shot 2020-05-26 at 12 18 24 PM" src="https://user-images.githubusercontent.com/1280564/82924955-14e2e880-9f4b-11ea-8f2f-8d7de0cae065.png">


2) Using `SIMPLIFIED_CATALOG_BASE=http://localhost:6500 npm run dev`:

Here it's also picking up the config file and saying it's running with the config file but it's actually not. I'm only able to access the "Edwin" library from the CM. So this runs correctly but the message is not right.
<img width="986" alt="Screen Shot 2020-05-26 at 12 21 25 PM" src="https://user-images.githubusercontent.com/1280564/82925461-bbc78480-9f4b-11ea-96aa-efb382d8378d.png">

Maybe it's that auth is not set up but when I click on "My books", I get direct to a page that says `You need to be signed in to view this page.` but there's no way to sign in.

It could be my configuration (I think this setup didn't work for me in the past anyway), but here's what I get for almost every book I click on (which wasn't a problem in the setup described in 1) ):

<img width="1067" alt="Screen Shot 2020-05-26 at 12 27 29 PM" src="https://user-images.githubusercontent.com/1280564/82925843-4dcf8d00-9f4c-11ea-892f-610d5ef33d95.png">

3) Using `CONFIG_FILE=config_file.txt npm run`:

This starts up fine - 
<img width="444" alt="Screen Shot 2020-05-26 at 12 29 15 PM" src="https://user-images.githubusercontent.com/1280564/82926015-8f603800-9f4c-11ea-8fef-b6005bc3f884.png">

Using the _any_ library in the config file to test, I get this on the homepage. This includes the `xyzlib` and `brazoria` libraries, as well as local libraries in my CM. Actually, this error pops up when I view individual books, too.

<img width="1273" alt="Screen Shot 2020-05-26 at 12 29 04 PM" src="https://user-images.githubusercontent.com/1280564/82926050-9e46ea80-9f4c-11ea-80eb-3786ff84b036.png">

----
### Production build

I get this error when trying to run `npm run build` (even after removing `node_modules` and installing again):
<img width="912" alt="Screen Shot 2020-05-26 at 12 47 39 PM" src="https://user-images.githubusercontent.com/1280564/82927719-20381300-9f4f-11ea-8912-01aa5ad297d3.png">

---- 
### Overall thoughts
I get the architecture change and I can't say I think it's any simpler just because I don't know Nextjs that well. I do think that clicking and transitioning is visibly slower. It's not the loading indicator but more so that when I click on a book or on the home button, there's a clear lag until anything seems to happen. This is also on the development instance and not the production build.

I'm hesitant about having the default jest test runner have `--watch` on. It's something that can be overlooked when deploying and forgetting to update the actions or Travis or whatever other CI tool is being used.

kristojorg commented:
Are 1+2 with the most recent code? Any way you can check after pulling? If you have a `.env` file or `.env.local` file set with eg a `CONFIG_FILE` and then add a `SIMPLIFIED_CATALOG_BASE` when running in the terminal, you'll end up having both set. At one point I had accidentally included a `CONFIG_FILE` setting in the `.env` file, which should be gone now.

For 1, I haven't been able to test at all with a library registry because I don't have one set up locally. Is there one deployed somewhere I can use?

If auth is not set up on your library you're testing with, the app isn't going to work right in general. With libraries I have auth set up in, the login modal does come up.

I fixed that `join` error - that was from having still not strict enough of typescript settings (we'll get there 😄 ) - aka I made a type error with an `any` type. 

On the production build - could you tell me what version of node you're using and possibly try once more with the most recent code?

**Overall thoughts**

The development instance does indeed have some lag when navigating. This is because it compiles pages lazily (when you request them), which apparently makes things like hot refresh etc faster and more stable. I find this also a bit annoying, but there is a little loading indicator in the bottom right, and after the first time you navigate to a page it should be pretty fast. It's basically just spreading out the long startup time we had before into chunks for each page. 

I think the ways that it'll be simpler haven't really been exposed yet. This just moves us over to Next.js without yet having taken advantage of really the core APIs it offers around server-rendering. That'll be down the line this summer. The two ways this specific PR does make things simpler is that it offloads the management of (a) the bundler, (b) the server, (c) the dev server/environment. It'll really start to shine later on though.

In terms of `jest --watch`, I did this because I would recommend always having it on while developing. I can definitely change it though if you prefer having `npm run jest` and `npm run jest:watch`

EdwinGuzman commented:
Re: `.env` file: I do have one but the environment variables are all commented out. I decided to run `npm run dev` with the env variable before it to make sure I knew what was running.

> For 1, I haven't been able to test at all with a library registry because I don't have one set up locally. Is there one deployed somewhere I can use?

I'm not sure there is one that is open.

I pulled the latest update and:
1. Using `REGISTRY_BASE=http://localhost:7000 npm run dev`:
same error as before

2. Using `SIMPLIFIED_CATALOG_BASE=http://localhost:6500 npm run dev`:
I'm not sure why my config is wrong. I get CORS errors even though my local CM should be open to any site. What I don't get is that the same "Edwin" library runs fine using step 1 from above, e.g. I cannot click on the ebook or audiobook icons using this step 2 but those buttons work fine using step 1.

3. Using `CONFIG_FILE=config_file.txt npm run`:
This is fine now.
The error I described in 1 also happens here (it was more apparent in _production_ mode) when I was on a book in the "Short Fiction" lane and then tried to click on "XYZ Public Library" in the breadcrumbs:
<img width="1207" alt="Screen Shot 2020-05-27 at 10 44 38 AM" src="https://user-images.githubusercontent.com/1280564/83035262-3b665980-a007-11ea-90f6-e80dfc887ad9.png">

This also showed up in the terminal when running in production mode but not sure how or why (maybe when I clicked on a book or the breadcrumbs):
<img width="1430" alt="Screen Shot 2020-05-27 at 10 47 09 AM" src="https://user-images.githubusercontent.com/1280564/83035507-8d0ee400-a007-11ea-9e1f-be0894da7dd5.png">

Notice the mismatch is happening in a missing slash in the second URL - `/xyzlib/book/URI/https:/standard ...`.

Production build:
I'm running it with node v10.19.0. This now works after the latest update.

kristojorg commented:
Okay I'll try to debug this this week. To summarize the errors, they are:
- LIBRARY_REGISTRY setting causes a url/as mismatch when clicking the breadcrumbs in your Edwin library
- eBook/audiobook icons can't be clicked on when running with SIMPLIFIED_CATALOG_BASE set to your local CM? Is that right?
- You get CORS errors using your local CM setup
- You get the same url mismatch in (more apparent in production) when running with CONFIG_FILE

Is this right? Sorry I got a bit confused by the description

EdwinGuzman commented:
Yes, that's right. Some more details:
* the LIBRARY_REGISTRY option breadcrumb issue happens with any library
* I think the second and third bullet points are related and probably just for my set up. @adriana-alter  if you have time and can test this PR branch, that'd be great.
* To be specific for the CORS issue, it doesn't happen in the CONFIG_FILE option, only the SIMPLIFIED_CATALOG_BASE option
* Yes.

adriana-alter commented:
Maybe it's my setup, but this has made it so I can't run the app at all 🙃

`REGISTRY_BASE` method produces this error (I tried it with several different libraries):
<img width="976" alt="Screen Shot 2020-05-27 at 1 00 38 PM" src="https://user-images.githubusercontent.com/42178216/83053477-c141cf00-a01e-11ea-87d1-511d33da1542.png">
 
`SIMPLIFIED_CATALOG_BASE` produces this error:
<img width="1080" alt="Screen Shot 2020-05-27 at 1 16 59 PM" src="https://user-images.githubusercontent.com/42178216/83053524-d9195300-a01e-11ea-9b32-080f8e9528e2.png">

`CONFIG_FILE` method doesn't run at all; just crashes in the terminal and seems to think the command is invalid.

kristojorg commented:
@adriana-alter have you removed `node_modules` and reinstalled? Could you tell me what version of node you are using?

@EdwinGuzman could you also tell me your node version? thanks

EdwinGuzman commented:
Node version is v10.19.0.

EdwinGuzman commented:
Wow, well the actions are pretty cool (in the latest commit):
<img width="1013" alt="Screen Shot 2020-05-29 at 9 46 22 AM" src="https://user-images.githubusercontent.com/1280564/83266837-6086d380-a191-11ea-8d89-7a0849c69fe6.png">

kristojorg commented:
Okay so here is what I have:
1. The href/as mismatch: I have fixed this error. It was a bug that sometimes we would get a collectionUrl from the backend which would just be the base url for the app, which I hadn't handled properly. I added a test for this in `Link.test.tsx`.
1. One of the mismatches you were experience - the one caused by a missing slash - I am not able to reproduce. Can you give me steps or something? Maybe even try it again now that I've fixed the first mismatch issue?
1. I am able to toggle with the audiobook/ebook filters in all of the libraries and configurations I have access to. Could you provide an example of a library where this doesn't work? How does it not work? Is there an error?
1. I'm also not able to reproduce the CORS issues. Is that specific to this PR?

EdwinGuzman commented:
All using the _same_ library:

1. Using `REGISTRY_BASE=http://localhost:7000 npm run dev`:
- Yes, the URL mismatch error is fixed now. 
- The message is the terminal about running with a library registry and a config file is also now fixed.

2. Using `SIMPLIFIED_CATALOG_BASE=http://localhost:6500 npm run dev`:
- The same library tested above returns 404 when attempting to view a book. It could be a CORS issue on my side but it's odd it happens in this config but not the `REGISTRY_BASE` config.
- The first error below is attempting to view a book.
- The next errors are when attempting to click on the audiobook/ebook tabs.

<img width="1097" alt="Screen Shot 2020-06-01 at 11 22 28 AM" src="https://user-images.githubusercontent.com/1280564/83424316-3dab2800-a3fa-11ea-9aba-dfda8ac74ebf.png">

3. Using `CONFIG_FILE=config_file.txt npm run dev`:
- This is working fine now with any library.
- The mismatch with the missing slash doesn't appear anymore.

------------------------------

- The CORS issue seems to just be my library.
- I noticed that when you build and run in production mode, the environment variables need to be set when building and not running. Can you make that more explicit in the readme? I should use the `.env` file but right now I'm just explicitly adding them before running any of the commands. Even when using an `.env` file, I think it makes sense to let the dev know when those environment variables will affect the app in the build process, i.e. if you run `CONFIG_FILE=config_file.txt npm run build` and then run `SIMPLIFIED_CATALOG_BASE=http://localhost:6500 npm run start`, the `SIMPLIFIED_CATALOG_BASE` environment variable will be ignored and only `CONFIG_FILE` will be used.

adriana-alter commented:
> @adriana-alter have you removed `node_modules` and reinstalled? Could you tell me what version of node you are using?

I did.  But, given that this is working for Edwin, I'm willing to believe that the issue is with my local setup rather than with your branch!  Node v11.12.0.

kristojorg commented:
- CORS Issue: This is interesting. I'm not sure what could be causing it to happen just with SIMPLIFIED_CATALOG_BASE. I'm also not sure what would cause that initial 404 error you got. Is this possibly because you should be running with `SIMPLIFIED_CATALOG_BASE=http://localhost:6500/edwin/groups` instead of just `localhost:6500`? Or, if that was just a typo, are you accidentally running it with an extra trailing slash in there?
- I will make a note about the env vars being set at build time in production

kristojorg commented:
Very odd, I'm not sure what I could have changed to surface that issue. I'll include it in the readme however. Where in the CM do you find that `/edwin/groups` url so I can point someone to look for the url of their specific library not the CM as a whole?
